models/classes/DomValidator.php

```diff
@@ -0,0 +1,210 @@
+            $this->errors = $errors ?? libxml_get_errors();
+        }
+
+        libxml_clear_errors();
+
+        return $isValid;
+    }
+
+    public function getErrors(): array
+    {
+        return $this->errors;
+    }
+
+    private function resetProperties(): void
+    {
+        $this->errors = [];
+        unset($this->assessmentItem, $this->validationKey, $this->hashedContent);
+    }
+
+    private function createDomDocument(): DOMDocument
+    {
+        $dom = new DomDocument();
+        $dom->formatOutput = true;
+        $dom->preserveWhiteSpace = false;
+
+        return $dom;
+    }
+
+    private function getAssessmentItem(DOMDocument $dom): ?DOMNode
+    {
+        return $dom->getElementsByTagName('assessmentItem')->item(0);
+    }
+
+    private function createValidationKey(string $schema): ?string
+    {
+        if ($this->assessmentItem !== null) {
+            $validationKey = md5($schema . $this->assessmentItem->getAttribute('identifier'));
+        }
+
+        return $validationKey ?? '';
+    }
+
+    /**
+     * When we are trying to update (save) item - we validate DOM content before processing and this content doesn't
+     * contain <responseProcessing/> node.
+     * When we are trying to preview item/booklet/etc. - we validate processed DOM content, that already contains it.
+     * In this case we need to remove this node to make stored content and validated content similar.
+     */
+    private function processAndHashContent(DOMDocument $dom): string
+    {
+        $responseProcessing = $dom->getElementsByTagName('responseProcessing')->item(0);
+
+        if ($responseProcessing !== null) {
+            $responseProcessing->parentNode->removeChild($responseProcessing);
+        }
+
+        return md5($dom->saveXML());
+    }
+
+    private function needValidation(): bool
+    {
+        return $this->assessmentItem === null
+            || !$this->hasValidatedContent($this->validationKey)
+            || $this->getValidatedContent($this->validationKey) !== $this->hashedContent;
+    }
+
+    private function hasValidatedContent(string $validationKey): bool
+    {
+        return $this->getKeyValuePersistence()->exists($validationKey . '.content');
+    }
+
+    private function getValidatedContent(string $validationKey): string
+    {
+        return (string) $this->getKeyValuePersistence()->get($validationKey . '.content');
+    }
+
+    private function getKeyValuePersistence(): KeyValuePersistence
+    {
+        if (!isset($this->keyValuePersistence)) {
+            $this->keyValuePersistence = $this->getOption(self::OPTION_KV_PERSISTENCE);
+
+            if ($this->keyValuePersistence === null || !$this->keyValuePersistence instanceof KeyValuePersistence) {
+                $this->keyValuePersistence = $this->getPersistenceManager()->getPersistenceById(
+                    'default_kv'
+                );
+            }
+        }
+
+        return $this->keyValuePersistence;
```

Comment:
It is safer to inject the persistence id instead of the object, cause on seed files it will not work :/ 

```suggestion
        if (!isset($this->keyValuePersistence)) {
            $this->keyValuePersistence = $this->getPersistenceManager()
                ->getPersistenceById($this->getOption(self::OPTION_KV_PERSISTENCE) ?? 'default_kv');
        }

        return $this->keyValuePersistence;
```